src/util/logging.cpp

```diff
@@ -43,41 +43,110 @@ Q_DECLARE_FLAGS(WriteFlags, WriteFlag)
 
 Q_DECLARE_OPERATORS_FOR_FLAGS(WriteFlags)
 
-// Handles actually writing to stderr and the log file.
+/// Format message for writing into log file (ignores QT_MESSAGE_PATTERN,
+/// because out logfiles should have a fixed format).
+inline QString formatLogFileMessage(
+        QtMsgType type,
+        const QMessageLogContext& context,
+        const QString& message,
+        const QString& threadName) {
+    Q_UNUSED(context);
+
+    QString levelName;
+    switch (type) {
+    case QtDebugMsg:
+        levelName = QStringLiteral("Debug");
+        break;
+    case QtInfoMsg:
+        levelName = QStringLiteral("Info");
+        break;
+    case QtWarningMsg:
+        levelName = QStringLiteral("Warning");
+        break;
+    case QtCriticalMsg:
+        levelName = QStringLiteral("Critical");
+        break;
+    case QtFatalMsg:
+        levelName = QStringLiteral("Fatal");
+        break;
+    }
+
+    return levelName + QStringLiteral(" [") + threadName + QStringLiteral("] ") + message;
+}
+
+/// Actually write a log message to a file.
+inline void writeToFile(
+        QtMsgType type,
+        const QMessageLogContext& context,
+        const QString& message,
+        const QString& threadName,
+        WriteFlags flags) {
+    DEBUG_ASSERT(flags & WriteFlag::File);
+    QMutexLocker locked(&s_mutexLogfile);
+    QString formattedMessageStr =
+            formatLogFileMessage(type, context, message, threadName) +
+            QChar('\n');
+    QByteArray formattedMessage = formattedMessageStr.toLocal8Bit();
+    // Writing to a closed QFile could cause an infinite recursive loop
+    // by logging to qWarning!
+    if (s_logfile.isOpen()) {
+        const int written = s_logfile.write(formattedMessage);
+        Q_UNUSED(written);
+        DEBUG_ASSERT(written == formattedMessage.size());
+        if (flags & WriteFlag::Flush) {
+            const bool flushed = s_logfile.flush();
+            Q_UNUSED(flushed);
+            DEBUG_ASSERT(flushed);
+        }
+    }
+}
+
+/// Actually write a log message to stderr.
+inline void writeToStdErr(
+        QtMsgType type,
+        const QMessageLogContext& context,
+        const QString& message,
+        const QString& threadName,
+        WriteFlags flags) {
```

Comment:
`WriteFlags flags` is redundant and confusing, we only need a boolean flush flag here.